### PR TITLE
OCPBUGS-16814: Use roundtripper for console client SA token renewal

### PR DESCRIPTION
**Pull request by jhadvig**

### CODE REVIEW

File: cmd/bridge/main.go [hunk 1]
```diff
@@ -377,25 +367,18 @@ func main() {
...
-		rootCAs := x509.NewCertPool()
-		if !rootCAs.AppendCertsFromPEM(k8sCertPEM) {
-			klog.Fatal("No CA found for the API server")
-		}
-		tlsConfig := oscrypto.SecureTLSConfig(&tls.Config{
-			RootCAs: rootCAs,
-		})
-
-		bearerToken, err := ioutil.ReadFile(k8sInClusterBearerToken)
+		tlsConfig, err := server.GetInClusterTLSConfig(caCertFilePath)
```
Comment (stlaz) [inline cmd/bridge/main.go:381]: unhandled error

File: cmd/bridge/main.go [hunk 2]
```diff
@@ -849,4 +779,9 @@ func main() {
 		klog.Info("not using TLS")
 		klog.Fatal(httpsrv.ListenAndServe())
 	}
+
+	if srv.K8sMode == "in-cluster" {
+		go srv.UpdateServiceAccountCertAndTokenPeriodically(oidcClientConfig, caCertFilePath, bearerTokenFilePath)
+		go srv.Test()
```
Comment (stlaz) [inline cmd/bridge/main.go:785]: left from debugging?

File: pkg/server/server.go [hunk 3]
```diff
@@ -988,3 +992,127 @@ func tokenToObjectName(token string) string {
...
+	}
+	return oscrypto.SecureTLSConfig(&tls.Config{
+		RootCAs: rootCAs,
+	}), nil
+}
+
+func (s *Server) Test() {
+	for {
+		time.Sleep(10 * time.Minute)
+		klog.Infof("\nTOKEN: %s\n", s.ServiceAccountToken)
```
Comment (stlaz) [inline pkg/server/server.go:1116]: in case this was not just for debugging purposes: you cannot log tokens

File: pkg/server/server.go [hunk 4]
```diff
@@ -988,3 +992,127 @@ func tokenToObjectName(token string) string {
...
+
+func (s *Server) UpdateClientConfigAuther(oidcClientConfig *auth.Config) {
+	var err error
+	if s.Authers[serverutils.LocalClusterName], err = auth.NewAuthenticator(context.Background(), oidcClientConfig); err != nil {
+		klog.Fatalf("Error updating authenticator: %v", err)
+	}
+}
+
+func GetInClusterToken(tokenPath string) (string, error) {
+	token, err := os.ReadFile(tokenPath)
```
Comment (stlaz) [inline pkg/server/server.go:1092]: consider using `transport.CachedFileTokenSource`

https://github.com/kubernetes/kubernetes/blob/7f532891c9c42500f75939fd931596cd42b1765b/staging/src/k8s.io/client-go/transport/token_source.go#L69-L70

File: pkg/server/server.go [hunk 5]
```diff
@@ -988,3 +992,127 @@ func tokenToObjectName(token string) string {
...
+	if err != nil {
+		return "", fmt.Errorf("error reading bearer token file: %v", err)
+	}
+	return string(token), nil
+}
+
+func GetInClusterTLSConfig(certPath string) (*tls.Config, error) {
```
Comment (stlaz) [inline pkg/server/server.go:1099]: consider using `dynamiccertificates.DynamicFileCAContent`

https://github.com/kubernetes/kubernetes/blob/6e0cb243d57592c917fe449dde20b0e246bc66be/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go#L76-L77

File: pkg/server/server.go (hunk 3 above)
Comment (jhadvig) [reply pkg/server/server.go:1116]: yep, just for debug

File: cmd/bridge/main.go (hunk 2 above)
Comment (jhadvig) [reply cmd/bridge/main.go:785]: yep, just for debug

File: pkg/server/server.go (hunk 4 above)
Comment (jhadvig) [reply pkg/server/server.go:995]: Used the `NewBearerAuthWithRefreshRoundTripper()` method in the `GetK8sClient()` function instead.

File: pkg/usersettings/handlers.go [hunk 6]
```diff
@@ -147,9 +146,8 @@ func (h *UserSettingsHandler) deleteUserSettings(ctx context.Context, client *ku
 
 func (h *UserSettingsHandler) createServiceAccountClient() (*kubernetes.Clientset, error) {
```
Comment (stlaz) [inline pkg/usersettings/handlers.go:147]: What's the reason to keep this function compared to simply initializing the client during the `UserSettingsHandler` struct initialization?

File: pkg/usage/metrics.go [hunk 7]
```diff
@@ -83,16 +82,14 @@ func (m *Metrics) MonitorUsers(
 func (m *Metrics) updateUsersMetric(
 	userSettingsClient *http.Client,
```
Comment (stlaz) [inline pkg/usage/metrics.go:83]: you may want to rename "userSettingsClient" to "saClient"

File: pkg/usage/metrics.go [hunk 8]
```diff
@@ -83,16 +82,14 @@ func (m *Metrics) MonitorUsers(
...
 
 	ctx := context.TODO()
 	config := &rest.Config{
-		Host:        userSettingsEndpoint,
-		BearerToken: serviceAccountToken,
-		Transport:   userSettingsClient.Transport,
+		Host:      userSettingsEndpoint,
+		Transport: userSettingsClient.Transport,
 	}
 	client, err := kubernetes.NewForConfig(config)
```
Comment (stlaz) [inline pkg/usage/metrics.go:94]: why don't you use a pre-created kubernetes SA clientset?

File: cmd/bridge/main.go [hunk 9]
```diff
@@ -566,11 +569,11 @@ func main() {
...
 	srv.KubeAPIServerURL = apiServerEndpoint
-	srv.K8sClient = &http.Client{
-		Transport: &http.Transport{
-			TLSClientConfig: srv.K8sProxyConfig.TLSClientConfig,
-		},
+	k8sClient, err := server.GetK8sClient(srv.K8sProxyConfig.TLSClientConfig, bearerTokenFilePath)
+	if err != nil {
+		klog.Fatalf("Error inferring Kubernetes config from environment: %v", err)
 	}
+	srv.K8sClient = k8sClient
```
Comment (stlaz) [inline cmd/bridge/main.go:576]: don't you need to specifically call the `RoundTrip()` method in order to use the enhanced roundtripper functionality?

File: cmd/bridge/main.go [hunk 10]
```diff
@@ -395,9 +398,9 @@ func main() {
 			RootCAs: rootCAs,
 		})
 
-		bearerToken, err := ioutil.ReadFile(k8sInClusterBearerToken)
+		bearerToken, err := server.GetInClusterToken(bearerTokenFilePath)
```
Comment (stlaz) [inline cmd/bridge/main.go:401]: why is it necessary to store the SA token string?

**Pull request closed without merging**